### DISCUSSION TITLE
Feedback on user expectation

### DISCUSSION THREAD

**Discussion by franburstall** [alphapapa/activities.el#64] [General]
This is less a bug report and more a brief account of what one user hoped to get from `activities`.  I set up an activity that I usually build by hand on startup containing 4 buffers: two `org-mode`, one `mastodon` and my `org` agenda.  Sadly the latter two do not support `bookmark` and so `activities-resume` failed (in a very helpful way that signposted the issue).

Anyway, the punchline is that the restriction to bookmark-able buffers will severely limit the usefulness of this package for me.

**Comment by alphapapa** [comment]
Hi Fran,

Thanks for the report.  First impressions are valuable feedback.

FWIW, `org-ql-view` buffers do support bookmarks, and thereby activities.el and burly.el, etc.  If you can rewrite your agenda buffer in terms of an `org-ql` query, that problem should be solved.

I can't help you with `mastodon`, but I'd guess that adding bookmark support to it shouldn't be hard, and it would be a good thing to add regardless of `activities`.

Anyway, `activities` can restore three kinds of buffers: 1) those that have `bookmark` support; 2) file-backed buffers; 3) buffers that are already open in Emacs and can be selected by name (like the `*scratch*` buffer).

And that is not so much a restriction as it is simply what is possible.  If a major mode isn't a simple file-backed mode, someone has to write code to save and restore its state.  And it wouldn't make sense to write support for individual major modes in this library when Emacs already has the bookmark system, which serves this purpose in a way that can be used across Emacs.

Having said that, a simple workaround might only take a few lines of code, e.g. something like this:
```elisp
(add-hook 'mastodon-mode
          (lambda ()
            (setq-local bookmark-make-record-function
                        (lambda ()
                          `((handler . (lambda (_bookmark)
                                         (call-interactively #'mastodon))))))))
```
That might suffice to bookmark the `mastodon` buffer in its default state.  All that matters generally is that the handler function returns with the current buffer being the desired one.

**Comment by franburstall** [comment]
>
> 3) buffers that are already open in Emacs and can be selected by name
> (like the *scratch* buffer).


Thanks for the clear explanation.  For my limited purpose, this solves the
problem: I fire `org-agenda` and `mastodon` to get their buffers open and
then do `activities-resume`.   I will also experiment with bookmarks given
your clear example: I wonder if something similar would work for org agenda
(or proced or....)

**Comment by franburstall** [comment]
> I will also experiment with bookmarks

This also gives a complete solution once I realised that handlers need to be symbols not lambdas.  Thus the following was enough for `activities` to save and restore my activity:

```el
;; make mastodon and org agenda bookmark-able to some degree
(defun my/agenda-bookmark (_bookmark)
  (org-agenda nil " "))

(defun my/mastodon-bookmark (_bookmark)
  (mastodon))

(add-hook 'org-agenda-mode-hook
          (lambda ()
            (setq-local bookmark-make-record-function
                        (lambda ()
                          `((handler . my/agenda-bookmark))))))

(add-hook 'mastodon-mode-hook
          (lambda ()
            (setq-local bookmark-make-record-function
                        (lambda ()
                          `((handler . my/mastodon-bookmark))))))
```
Clearly I should write a macro if I was going to do this more often!

Anyway, feel free to close the ticket.  Thanks for the advice (and the nice package).

**Reply by alphapapa** [reply]
> This also gives a complete solution once I realised that handlers need to be symbols not lambdas.

AFAIK that's not the case; I think the requirement is that the bookmark record be printable and readable.  Lambdas are both, but if one of them bound or was a closure around an unprintable/unreadable value...

> Clearly I should write a macro if I was going to do this more often!

Even better, contribute those functions upstream.  :)  Even if they are basic initially, they may inspire others to fill them out with more functionality.

>  Thanks for the advice (and the nice package).

Thanks.

**Comment by martianhi** [comment]
hi all, 

i maintain mastodon.el. is the above code enough to support bookmarks in mastodon.el generally or is it specific to this library? mastodon.el doesn't have bookmark support mainly because i don't know how to implement it and no one else has either.

(it's also not clear to me how to bookmark something as dynamic as a masto buffer. i'm totally in the dark.)

thanks!

**Reply by jdtsmith** [reply]
I think this workaround simply runs `(mastodon)` for any saved mastodon buffers in an activity.  A better approach would be to write a function which returns some metadata in the `bookmark-make-record-function` callback.  This would encapsulate the current mastodon state.  Then you'd also provide in the returned alist the name of a mastodon custom handler function which can restore that state.  When a bookmark is loaded, your handler function gets called with the bookmark, which can be queried for the data you squirreled away in it.  

These can be quite trivial functions.  You can check `activities-bookmark-handler` or `Info-bookmark-jump` for examples of that approach being used.

**Reply by alphapapa** [reply]
@mooseyboots Yes, what JD said.  :)  Assuming you have a function which can be called with, e.g. a URL or something, you just need to make a bookmark record (which is just an alist) with that value and a pointer to the handler function.  Then the handler will call that function with the value and recreate the buffer as if the user ran the command.  The most important thing to keep in mind is that the "bookmarked" buffer must be the current buffer when the handler function returns (i.e. don't wrap the body in `with-current-buffer`).

**Comment by martianhi** [comment]
thanks. I was motivated to add basic bookmark handling to mastodon.el due to the above discussion and code. glad I learned how it works now. as you mentioned, it can start out very basic, and add handling of specific cases as needed.

**Reply by ParetoOptimalDev** [reply]
Great! I'm looking forward to it.
